codec-http2/src/main/java/io/netty/handler/codec/http2/StreamBufferingEncoder.java

```diff
@@ -0,0 +1,347 @@
+import static io.netty.handler.codec.http2.Http2Error.PROTOCOL_ERROR;
+import static io.netty.handler.codec.http2.Http2Exception.connectionError;
+
+import io.netty.buffer.ByteBuf;
+import io.netty.channel.ChannelFuture;
+import io.netty.channel.ChannelHandlerContext;
+import io.netty.channel.ChannelPromise;
+import io.netty.util.ReferenceCountUtil;
+
+import java.util.ArrayDeque;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.Queue;
+import java.util.TreeMap;
+
+/**
+ * Implementation of a {@link Http2ConnectionEncoder} that dispatches all method call to another
+ * {@link Http2ConnectionEncoder}, until {@code SETTINGS_MAX_CONCURRENT_STREAMS} is reached.
+ * <p/>
+ * <p>When this limit is hit, instead of rejecting any new streams this implementation buffers newly
+ * created streams and their corresponding frames. Once an active stream gets closed or the maximum
+ * number of concurrent streams is increased, this encoder will automatically try to empty its
+ * buffer and create as many new streams as possible.
+ * <p/>
+ * <p>
+ * If a {@code GOAWAY} frame is received from the remote endpoint, all buffered writes for streams
+ * with an ID less than the specified {@code lastStreamId} will immediately fail with a
+ * {@link StreamBufferingEncoder.GoAwayException}.
+ * <p/>
+ * <p>This implementation makes the buffering mostly transparent and is expected to be used as a
+ * drop-in decorator of {@link io.netty.handler.codec.http2.DefaultHttp2ConnectionEncoder}.
+ */
+public class StreamBufferingEncoder extends DecoratingHttp2ConnectionEncoder {
+
+    /**
+     * Thrown by {@link StreamBufferingEncoder} if buffered streams are terminated due to
+     * receipt of a {@code GOAWAY}.
+     */
+    public static final class GoAwayException extends Http2Exception {
+        private static final long serialVersionUID = 1326785622777291198L;
+        private final int lastStreamId;
+        private final long errorCode;
+        private final ByteBuf debugData;
+
+        private GoAwayException(int lastStreamId, long errorCode, ByteBuf debugData) {
+            super(Http2Error.STREAM_CLOSED);
+            this.lastStreamId = lastStreamId;
+            this.errorCode = errorCode;
+            this.debugData = debugData;
+        }
+
+        public int lastStreamId() {
+            return lastStreamId;
+        }
+
+        public long errorCode() {
+            return errorCode;
+        }
+
+        public ByteBuf debugData() {
+            return debugData;
+        }
+    }
+
+    /**
+     * Buffer for any streams and corresponding frames that could not be created due to the maximum
+     * concurrent stream limit being hit.
+     */
+    private final TreeMap<Integer, PendingStream> pendingStreams =
+            new TreeMap<Integer, PendingStream>();
+    private final int initialMaxConcurrentStreams;
+    // Smallest stream id whose corresponding frames do not get buffered.
```

Comment:
@buchgr - Agreed it is correct. Wondering if it may be more clear to not have the comment use opposite verbiage as the variable name.  Something like:

```java
/**
 * Largest stream ID that is not buffered in {@link pendingStreams}.
 */
```